dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/datafile/page/WholeDatasetDownloadUiLogger.java

```diff
@@ -0,0 +1,150 @@
+            return;
+        }
+
+        // In following "if" the order of checking is important
+        if (isUserDownloadingAllFilesFromCurrentVersion(filesMetadata)
+            || isUserDownloadingAllFilesFromAnyVersion(filesMetadata)) {
+            downloadDatasetLogService.incrementDownloadCountForDataset(datasetId);
+        }
+    }
+
+    // -------------------- PRIVATE --------------------
+
+    /**
+     * Because from the UI we can only download files from the same version, it's sufficient
+     * that we only compare the size of list of files being downloaded to the size of list of files
+     * in current version
+     */
+    private boolean isUserDownloadingAllFilesFromCurrentVersion(List<FileMetadata> filesMetadata) {
+        long currentVersionId = extractVersionIdOfDownloadedFiles(filesMetadata);
+        int currentVersionFileCount = extractCurrentVersionFileCountFromMetadata(currentVersionId, filesMetadata);
+
+        return filesMetadata.size() == currentVersionFileCount;
+    }
+
+    /**
+     * Here we are extracting sets of ids of files from all the versions that have no more files than
+     * list of files being downloaded and we check if any of these sets is wholly contained in the set of ids
+     * of files being downloaded
+     */
+    private boolean isUserDownloadingAllFilesFromAnyVersion(List<FileMetadata> filesMetadata) {
+        Set<Long> fileIdSet = mapFileMetadataToIds(filesMetadata);
+        List<Set<Long>> fileIdSetsInActiveVersions = createPossibleSetsOfFileIdsFromVersions(filesMetadata, fileIdSet);
+
+        return fileIdSetsInActiveVersions.stream()
+                .anyMatch(fileIdSet::containsAll);
+    }
+
+    private Optional<FileMetadata> takeAnyFileMetadata(List<FileMetadata> filesMetadata) {
+        return filesMetadata.stream()
+                .findFirst();
+    }
+
+    private Long extractVersionIdOfDownloadedFiles(List<FileMetadata> filesMetadata) {
+        return takeAnyFileMetadata(filesMetadata)
+                .map(FileMetadata::getDatasetVersion)
+                .map(DatasetVersion::getId)
+                .orElse(-1L);
+    }
+
+    private Optional<Dataset> extractDataset(List<FileMetadata> filesMetadata) {
+        return takeAnyFileMetadata(filesMetadata)
+                .map(FileMetadata::getDataFile)
+                .map(DataFile::getOwner);
+    }
+
+    private Long extractDatasetId(List<FileMetadata> filesMetadata) {
+        return extractDataset(filesMetadata)
+                .map(Dataset::getId)
+                .orElse(-1L);
+    }
+
+    private Integer extractCurrentVersionFileCountFromMetadata(long currentVersionId, List<FileMetadata> filesMetadata) {
+        return extractDataset(filesMetadata)
+                .map(Dataset::getVersions)
+                .orElseGet(Collections::emptyList).stream()
+                .filter(v -> v != null && Long.valueOf(currentVersionId).equals(v.getId()))
```

Comment:
To be perfectly honest I have a totally different view of this issue: in case of lambda we have the same name of variable that is used at least 2 times, and (in majority of cases) in the scope of one single line of code, so IMHO it seems a bit noisy.

Reply:
Yeah I agree that's different situation since mapping is above it and you can deduct it, but for me it's just more readable to have full/shortcut name. I think we can let it be agree to disagree and you can just leave it as it is.